src/librustc_mir/interpret/intern.rs

```diff
@@ -328,10 +336,29 @@ pub fn intern_const_alloc_recursive<M: CompileTimeMachine<'mir, 'tcx>>(
             // We can't call the `intern_shallow` method here, as its logic is tailored to safe
             // references and a `leftover_allocations` set (where we only have a todo-list here).
             // So we hand-roll the interning logic here again.
-            match base_intern_mode {
-                InternMode::Static => {}
-                InternMode::Const | InternMode::ConstBase => {
-                    // If it's not a static, it *must* be immutable.
+            match intern_kind {
+                // Mutable statics may contain mutable allocations even behind relocations
+                InternKind::Static(hir::Mutability::Mut) => {}
+                // Once we get heap allocations we need to revisit whether immutable statics can
+                // refer to mutable (e.g. via interior mutability) allocations.
+                // Note: this is never the base value of the static, we can only get here for
+                // pointers encountered inside the base allocation, and then only for ones not at
+                // reference type, as that is checked by the type based main interner.
+                InternKind::Static(hir::Mutability::Not) => {
+                    alloc.mutability = Mutability::Not;
+                }
+                // Raw pointers in promoteds may only point to immutable things so we mark
+                // everything as immutable. Creating a promoted with interior mutability is UB, but
+                // there's no way we can check whether the user is using raw pointers correctly.
+                // So all we can do is mark this as immutable here.
+                // It is UB to mutate through a raw pointer obtained via an immutable reference.
+                // Since all references and pointers inside a promoted must by their very definition
+                // be created from an immutable reference, mutating though them would be UB.
+                InternKind::Promoted => {
+                    alloc.mutability = Mutability::Not;
```

Comment:
Okay, so this is basically reverting what you did when you added the `delay_span_bug`: turns out that code is basically unreachable, and in the reachable variant we have here, everything is unchanged. Does that make sense?

Reply:
yes

Reply:
All right, this new branch is fine for me then (modulo the comment). That's the critical part, I think.